pkg/util/runtimeconfig/manager.go

```diff
@@ -131,7 +139,21 @@ func (om *Manager) loop(ctx context.Context) error {
 // loadConfig loads configuration using the loader function, and if successful,
 // stores it as current configuration and notifies listeners.
 func (om *Manager) loadConfig() error {
-	cfg, err := om.cfg.Loader(om.cfg.LoadPath)
+	var reader io.Reader
+	var hash []byte
+
+	if om.cfg.LoadPath != "" {
+		buf, err := ioutil.ReadFile(om.cfg.LoadPath)
+		if err != nil {
+			om.configLoadSuccess.Set(0)
+			return err
+		}
+		reader = bytes.NewReader(buf)
+		hashSum := sha256.Sum256(buf)
+		hash = hashSum[:]
+	}
+
+	cfg, err := om.cfg.Loader(reader)
```

Comment:
This will call loader even if there was no file, `reader` will be nil. I think that's not desirable.

Reply:
Maybe we should document a bit more how the loader method should behave, as right now there is not really a comment on. Maybe it needs to deal with a nil reader?

 I basically added that check, so I could make those two test cases work where there is no file provided (`TestOverridesManager_ListenerChannel` and `TestOverridesManager_StopClosesListenerChannels`)

Without a path this whole module wouldn't even be initialised (I think): https://github.com/cortexproject/cortex/blob/981bf9df1b3d8524a593e73405d6c057328ecc76/pkg/cortex/modules.go#L131

What would be a good approach? Rewrite those two test cases to use files instead?

Reply:
> What would be a good approach? Rewrite those two test cases to use files instead?

Previously Manager wasn't involved in path check, now it is since it opens the file on its own. Tests need to be modified to actually use files now.

I think we should `NewRuntimeConfigManager` to return error if path is empty. Then there is no `nil` reader problem. Since we don't even try to initialize the Manager it in such case from the rest of Cortex, this seems to be easy way out. (Note that Loki may still be doing that... it should be fixed afterwards).

Reply:
I agree. We shouldn't workaround the code to make tests pass, but improve tests. All tests should load config from file. You can generate a temporary file via code (with deferred deletion) and then load it in the test.